### ISSUE TITLE
Multiple queues

### ISSUE THREAD

**Issue by BoPeng** [vatlab/SoS#1063]
Currently sos supports one queue for each task. The queue can be a single machine or a cluster. However, in a model similar to ipyparallel, sos could  send tasks to multiple queues or hosts. Handling this situation is a lot harder because we do not have workers on remote hosts so load balancing etc will be challenging.  Just throw the ticket here as something to consider later.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BoPeng** [comment]
Sorry, I created this ticket from my cellphone without details. I have updated the ticket.

**Comment by gaow** [comment]
I see, of course. Let me try to interpret the implications. One benefit implementing the `ipyparallel` model would be more balanced workload DAG-wise, not locally. Another would be a more homogeneous job controlling mechanism and allows the potential to natively support more systems (like nextflow?), although it might be hard to reach to a level of sophistication of a system like PBS. Right?

**Comment by BoPeng** [comment]
Yes, the "mini-cluster" approach of nextflow means it has some built-in load balancing features and we generally avoids this problem by letting others to manage the execution of large amount of tasks.

**Comment by BoPeng** [comment]
Our model currently is

1. Keep DAG and light computation (steps) locally, even with multiprocessing.
2. Send specialized tasks to remote hosts as processes.
3. Send heavy computations to task queues.

Nexflow's submit multi-node jobs on clusters and execute everything in the mini-cluster, so it also does heavy computations in nextflow. SoS could also do this by allowing our workers to reside on multiple computing nodes and converting them also to workers that handle jobs (substeps). We have already converted from local multiprocessing to single node zmq messaging so expanding the architecture to remote zmq workers would allow us to do something like that. That is to say, with something like

```
mpirun -n8 sos run  script -m
```
we can start sos on 8 nodes and one sos instance can talk to other sos instances as workers (instead of creating sos workers locally). We can ignore option `-q` (or `task: queue=""`) in this case and submit all tasks as concurrent jobs to workers. This can potentially be much more efficient especially for your case (large amount of small tasks/jobs).
